### PR TITLE
Propagate committed subdag

### PR DESCRIPTION
**Pull request by asonnino**
Before this PR, consensus shipped committed certificates one by one through an output channel. It now sends the entire committed sub-dag to the executor.

The consensus now ships the following structure to the executor: https://github.com/MystenLabs/sui/blob/bee6ab6f712d29ff6759fdf11365178f71aa0344/narwhal/consensus/src/lib.rs#L38-L43

Then the executor performs the following tasks:
* The [subscriber](https://github.com/MystenLabs/sui/blob/main/narwhal/executor/src/subscriber.rs) looks at this structure to fetch a list of batches (i.e., the transaction data) from the workers.
* The subscriber then ships that list to the [notifier](https://github.com/MystenLabs/sui/blob/main/narwhal/executor/src/notifier.rs) that executes them. The notifier is also augmented to know when each sub-dag has been committed (so it can call `notify_commit_boundary`)
 
To not break the interface with Sui, this PR does not allow the `Authority` to actually make use of the sub-dag yet; this will be the job of the next PR.

Further thoughts:
* Is there still a need to store the entire committed sequence at the consensus layer?
* Should we simplify execution indices?

### CODE REVIEW

File: narwhal/consensus/src/consensus.rs
Comment (andll) [inline narwhal/consensus/src/consensus.rs:217]: I don't think we actually need `Box` here - note that `std::Vec` already puts all it's content on the heap, and with this in mind `CommittedSubDag` is not a very large data structure as it only contains Certificate + Vec(and we seem to be ok to sending Certificate across the channels)

**Review by andll — CHANGES_REQUESTED** [review]
Left some comments, my main question is that I don't fully understand why `self.committee.leader(round) == self.name` denote a commit boundary

File: narwhal/executor/src/notifier.rs
Comment (akichidis) [inline narwhal/executor/src/notifier.rs:76]: I believe this comment now needs to change? https://github.com/MystenLabs/sui/blob/e6f8cee235365d4723e16234e8ff811ea2182e98/narwhal/executor/src/lib.rs#L57

Comment (akichidis) [reply narwhal/executor/src/notifier.rs:76]: I am also a little bit curious why we needed this extra callback here. The way we seem to use this really follows the exact way we execute things. I was thinking whether we could just piggy back the `ConsensusOutput` to mark the commit boundaries. Probably outside of the scope of this PR but since I am looking how this code evolves I would suggest _maybe_ rethinking the approach - unless there is some further design which I ignore here.

File: narwhal/executor/src/notifier.rs
Comment (akichidis) [reply narwhal/executor/src/notifier.rs:74]: Actually it looks like we'll notify the callback on each batch execution for the corresponding certificate - so it won't execute once.

I would suggest getting this signal from the upstream component instead of making `notifier` inferring the commit boundary by using the leader round as the `executor` and the `consensus` are two different components. In this case I would suggest either:
* having a `bool` value in the `index` that would suggest that this `index` is a commit boundary
* consume from the `rx_notifier` channel a tuple of `Some((index, batch, commit_boundary))`

File: narwhal/executor/src/notifier.rs
Comment (andll) [reply narwhal/executor/src/notifier.rs:76]: We do need a callback because sui needs to know when the commit boundary is reached (we actually already have an impl for it in main, so it can provide some context on how it is used).

In terms of how this callback can be implemented I thought about two approaches:

(1) As you say we can perhaps just add a bool to ConsensusOutput (smth like `is_last_in_commit`)
(2) What I thought can be cleaner if we just re-think what consensus output is - instead of sequencing certificates, we sequence Commits. `ConsensusOutput::index` becomes an index of commit rather then height of certificate, `ConsensusOutput::content` becomes `Vec<Certificate>`, instead of `Certificate`. And ExecutionIndexes becomes a 4-touple instead of 3-touple, e.g. holds `(commit_index, certificate_index_in_commit, batch_index_in_certificate, transaction_index_in_batch)`.

I believe that approach (2) is more flexible in terms of what can come next (e.g. for example for MEV would need entire commit in the executor)

File: narwhal/executor/src/notifier.rs
Comment (andll) [reply narwhal/executor/src/notifier.rs:74]: Replied below - while I think bool approach would work in theory, I think passing entire commit from consensus to executor can unlock more possibilities in the future

**Comment by asonnino** [comment]
I see, so we want to pass the committed sub-dag to the notifier (not to the execution crate in general)

**Comment by asonnino** [comment]
Ok, have a look now if it is what you had in mind. The `notifier` has now access to the committed sub-dag

File: narwhal/executor/src/notifier.rs
Comment (gdanezis) [inline narwhal/executor/src/notifier.rs:62]: Here I would change the interface to pass in the whole commit. And allow Sui to process the whole commit as a whole. There is no point feeding Sui or anyone else a commit at a time, and having to manage this process if it gets interrupted half way through.

File: narwhal/executor/src/notifier.rs
Comment (gdanezis) [inline narwhal/executor/src/notifier.rs:67]: As per above, why do we feed things one by one, and then require this extra thing here to denote the last. Presumably the other side now will be re-constructing the vector, and waiting for this call for processing. Lets just send the whole commit and avoid the complexity?

File: narwhal/executor/src/notifier.rs
Comment (asonnino) [reply narwhal/executor/src/notifier.rs:62]: totally agree, I planned to do it in a separate PR to keep things short

File: narwhal/executor/src/notifier.rs
Comment (asonnino) [reply narwhal/executor/src/notifier.rs:67]: yes, I wanted to avoid touching too much at the SUI codebase in a single PR

Comment (asonnino) [reply narwhal/executor/src/notifier.rs:67]: I should have made that clear in the PR description (I now realise)

File: narwhal/node/src/lib.rs
Comment (andll) [inline narwhal/node/src/lib.rs:324]: Let's remove commented code?

**Review by andll — COMMENTED** [review]
Thanks, this now seem more like what we want :)
Left some more comments.

I also have a question about wrapping/unwrapping subdag into shell - why did you chose to do it this way? Would it be easier/less error prone if we just had one CF for entire subdag?

File: narwhal/executor/src/state.rs
Comment (asonnino) [reply narwhal/executor/src/state.rs:17]: Good idea. Are we sure we need to implement Ord? The current order is correct (although I guess it will make things cleaner)

File: narwhal/executor/src/notifier.rs
Comment (asonnino) [reply narwhal/executor/src/notifier.rs:66]: ha good point

File: narwhal/executor/tests/consensus_integration_tests.rs
Comment (akichidis) [inline narwhal/executor/tests/consensus_integration_tests.rs:138]: I would suggest changing this to :

```
        assert_eq!(consensus_output[0].len(),  consensus_index_counter as usize
```

and we probably need to refactor the test (maybe in follow up PR) to better capture the current recovery semantics.

**Before this PR**, we were effectively recovering _all the committed certificates_ starting from _the last executed certificate_

**After this PR** we'll recover _all the committed certificates_ since the last committed subdag/leader. So we'll always expect to replay at least all the certificates of the latest subdag.

as we discussed offline nothing wrong given SUI's idempotency , but we might want to make it clearer in the test it self

File: narwhal/node/src/lib.rs
Comment (akichidis) [inline narwhal/node/src/lib.rs:317]: Github doesn't allow me to commend on a non-changed line, but let's change the `info!` message at line 328
https://github.com/MystenLabs/sui/pull/5900/files#diff-f6d9e991182542565a5ad10f44f1eeeb4a9536bcae12648c36be4a1f47a12ab0R328

to accurately capture the restored quantities. For example it could be something like:

```
Consensus output on its way to the executor was restored for {} leaders and {} certificates"
```

File: narwhal/types/src/consensus.rs
Comment (akichidis) [inline narwhal/types/src/consensus.rs:33]: I am thinking whether we actually need this stored here or we can infer from the `certificates` anyways. I mean, we expect the `certificates` to be ordered and the leader should always be the last "committed"  - so we expect it to be the last certificate of the list. So practically since the DAG is linearised I am thinking whether we would be equally ok of just storing the `certificates` and on the fly get the leader when needed. Then the struct could be reduced to just:

```
struct CommittedSubDag(Vec<ConsensusOutput>)
```

and could have a method like:

```
    pub fn leader(&self) -> Option<ConsensusOutput> {
        self.0.last().cloned()
    }
```

**Review by akichidis — COMMENTED** [review]
Thanks for the PR @asonnino !! I've left some comments

File: narwhal/types/src/consensus.rs
Comment (asonnino) [reply narwhal/types/src/consensus.rs:33]: Are we sure the leader will be the last committed? I see that we have (scary) modular code to traverse the dag now. I was afraid that we may change it in the future and break safety. Shouldn't we instead try to remove the list of committed certificates from storage?

File: narwhal/consensus/src/bullshark.rs
Comment (asonnino) [reply narwhal/consensus/src/bullshark.rs:193]: It can be arbitrarily deep (depending on the network adversary)

Comment (akichidis) [reply narwhal/consensus/src/bullshark.rs:193]: We also talked offline about this. Having this unbounded in size is a bit concerning as we can't really predict the network characteristics. I would suggesting leaving a comment/TODO here as a point to keep in mind.

File: narwhal/types/src/consensus.rs
Comment (akichidis) [inline narwhal/types/src/consensus.rs:135]: Could you please add some instrumentation here? (maybe in `trace` level). I would be curious to keep track of the performance of this.

```
#[instrument(level = "trace", skip_all)]
```

File: narwhal/executor/src/notifier.rs
Comment (akichidis) [reply narwhal/executor/src/notifier.rs:69]: I also assume we are ok with the "at least once" semantics of calling the `notify_commit_boundary` method right? I mean during crash recovery this will be called again in case of commit replays

Comment (akichidis) [reply narwhal/executor/src/notifier.rs:69]: Also looking on the increased responsibility of the `notifier` I would suggest adding some test coverage (even just the happy path)

**Comment by akichidis** [comment]
> I also have a question about wrapping/unwrapping subdag into shell - why did you chose to do it this way? Would it be easier/less error prone if we just had one CF for entire subdag?

@andll could you please elaborate more on this idea? Wanna make sure I get it right before I comment

File: narwhal/types/src/consensus.rs
Comment (asonnino) [reply narwhal/types/src/consensus.rs:33]: > > I also have a question about wrapping/unwrapping subdag into shell - why did you chose to do it this way? Would it be easier/less error prone if we just had one CF for entire subdag?
> 
> @andll could you please elaborate more on this idea? Wanna make sure I get it right before I comment

Oh I missed this comment somehow. I don't like the sub-dag shell, but it seemed wrong to persist the entire certificate's data. So we only persist digests of the certificates and retrieve the corresponding data upon crash-recovery.

**Comment by akichidis** [comment]
@asonnino don't have any other comments from my side. My only comment is whether you want to deploy this in one of our environments (ex private testnet or the labnet) for some testing before you merge - or you are good to go as is.

**Comment by asonnino** [comment]
> @asonnino don't have any other comments from my side. My only comment is whether you want to deploy this in one of our environments (ex private testnet or the labnet) for some testing before you merge - or you are good to go as is.

However you prefer (no strong feelings). Although It seems that I cannot merge it since there still seems to be an issue with the Typescript SDK

**Comment by akichidis** [comment]
> However you prefer (no strong feelings). Although It seems that I cannot merge it since there still seems to be an issue with the Typescript SDK

Yeah that's a known issue for now - and not blocking for merging. Also you don't need to merge to deploy in any of the testing environments - we can just deploy from any commit. Ping me for doing this offline. I am disabling auto-merge and approving to unblock. Then you do as you wish.

**Review by akichidis — APPROVED** [review]

**Comment by asonnino** [comment]
@andll it seems we also need your approval to allow merging

**Review by andll — APPROVED** [review]

**Pull request merged**
